boot/modeenv.go

```diff
@@ -44,7 +44,11 @@ type Modeenv struct {
 	read bool
 }
 
-func ReadModeenv(rootdir string) (*Modeenv, error) {
+var ReadModeenv = readModeenv
```

Comment:
(nitpick^99) A docstring for ReadModeenv might be nice so that we have some doc in `go doc boot` for  it. Right now the `go doc boot` output for it is not super nice because it's grouped under variables. Maybe (and I think Samuele should comment if that makes sense or not) something like:
```
diff --git a/boot/boottest/modeenv.go b/boot/boottest/modeenv.go
index cb61f7fee0..4f90cca7e3 100644
--- a/boot/boottest/modeenv.go
+++ b/boot/boottest/modeenv.go
@@ -29,17 +29,13 @@ import (
 // given root dir, returning a restore function to reset to the old behavior.
 // If rootdir is empty, then all invocations return the specified modeenv
 func ForceModeenv(rootdir string, m *boot.Modeenv) func() {
-       old := boot.ReadModeenv
-       boot.ReadModeenv = func(callerrootdir string) (*boot.Modeenv, error) {
+       restore := boot.MockReadModeenv(func(callerrootdir string) (*boot.Modeenv, error) {
                if rootdir == "" || callerrootdir == rootdir {
                        return m, nil
                }
 
                // all other cases return doesn't exist
                return nil, os.ErrNotExist
-       }
-
-       return func() {
-               boot.ReadModeenv = old
-       }
+       })
+       return restore
 }
diff --git a/boot/modeenv.go b/boot/modeenv.go
index 439788723d..7bb45ed105 100644
--- a/boot/modeenv.go
+++ b/boot/modeenv.go
@@ -44,11 +44,19 @@ type Modeenv struct {
        read bool
 }
 
-var ReadModeenv = readModeenv
-
-// readModeenv will attempt to read the file specified by
+// ReadModeenv will attempt to read the file specified by
 // "rootdir/dirs.SnapModeenvFile" as a Modeenv
-func readModeenv(rootdir string) (*Modeenv, error) {
+func ReadModeenv(rootdir string) (*Modeenv, error) {
+       return readModeenvImpl(rootdir)
+}
+
+func MockReadModeenv(f func(rootdir string) (*Modeenv, error)) (restore func()) {
+       old := readModeenvImpl
+       readModeenvImpl = f
+       return func() { readModeenvImpl = old }
+}
+
+var readModeenvImpl = func(rootdir string) (*Modeenv, error) {
        modeenvPath := filepath.Join(rootdir, dirs.SnapModeenvFile)
        cfg := goconfigparser.New()
        cfg.AllowNoSectionHeader = true
``` 
Or even have ForceModeenv already part of boot instead of boottest.

Anyway just an observation while reading this, not a blocker and we can always iterate on this later.

Reply:
IIRC, we use this pattern in snapd-update-ns too. I'll push an update here.